### ISSUE TITLE
Warn about `$x:expr` in `macro_rules` expansion

### ISSUE THREAD

**Issue by jorendorff** [rust-lang/rust#39822]
Maybe I'm the only one, but a stupid mistake I've made several times, when writing macros, is to accidentally include `$x:expr` in the output of a macro, when of course I should just write `$x`.

The error messages when I do this are correct... but there's so much going on that I *still* typically don't understand until I've spent a few minutes enabling `trace_macros!` and poring over the output.

It should be possible to warn about this when the macro is defined. I think it's always a mistake.

**Comment by durka** [comment]
It's not a mistake if you're declaring a nested macro, which is an exceedingly rare scenario. I've bitten myself with this frequently as well, usually with `$($x:tt)*`.

**Comment by jorendorff** [comment]
True, and yet it seems like it'd be a feature, not a bug, to have to `#[allow(suspicious_macro_output)]` when writing a nested macro like that :)

(It'd be different if the system were designed to support nested macros. But that doesn't seem to be the case—not being able to quote `$` has thwarted me every time I've tried.)
